src/Disks/ObjectStorages/MetadataStorageFromPlainRewritableObjectStorage.cpp

```diff
@@ -45,6 +52,38 @@ std::string getMetadataKeyPrefix(ObjectStoragePtr object_storage)
         : metadata_key_prefix;
 }
 
+void loadDirectoryTree(
+    InMemoryDirectoryPathMap::Map & map, InMemoryDirectoryPathMap::FileNames & unique_filenames, ObjectStoragePtr object_storage)
+{
+    using FileNamesIterator = InMemoryDirectoryPathMap::FileNamesIterator;
+    using FileNameIteratorComparator = InMemoryDirectoryPathMap::FileNameIteratorComparator;
+    const auto common_key_prefix = object_storage->getCommonKeyPrefix();
+    LOG_DEBUG(getLogger("MetadataStorageFromPlainObjectStorage"), "Loading directory structure");
+    for (auto & [local_path, info] : map)
+    {
+        const auto remote_path = std::filesystem::path(common_key_prefix) / info.path / "";
+        std::set<FileNamesIterator, FileNameIteratorComparator> filename_iterators;
+        for (auto iterator = object_storage->iterate(remote_path, 0); iterator->isValid(); iterator->next())
```

Comment:
Just an idea, not sure if it works: 

UPDATE: I just remembered that we do actually have quite flat layout of blobs in object storage, so the complexity of this method should be small, so most likely the idea I had is not worth it.

the idea:
it seems it is possible to avoid traversing nested directories several times and thus reduce complexity of this method, so would it be better if we constructed a temporary map just in this method according to `map` we already have (`local_path` -> `info`) to have a temporary map: `remote_path` -> `map_it` (`map_it` from the map we already had, `remote_path` is `std::filesystem::path(common_key_prefix) / info.path / ""`). So then it seems to be enough to just traverse only root directories (if we do not have a single root directory) which would give as all its inner files recursively and then just check that if `auto it = tmp_map.find(iterator->current().parent_path())` then insert `iterator->current().filename()` as `it->second.info.filename_iterators.emplace(iterator->current().filename())`. Does this sound plausible?

Reply:
We have flat structure starting from 24.8, and hierarchical t prior to 24.8. Older tables should be removed from cloud starting from recently.

> traverse only root directory 

I did not benchmark, but suspect that traversing all files in one thread might be slow. The initial loading was pretty slow when metadata 'prefix.path' files was stored along with the regular MergeTree files, and we saw improvement after moving metadata files into `__meta`:
https://github.com/ClickHouse/ClickHouse/issues/29693#issuecomment-2274054072
I'm adding `ThreadPoolCallbackRunnerLocal` to parallelize the `loadDirectoryTree` function.

Reply:
Looks like some implementations of `iterateDirectory` are not thread-safe? I'll revert the last commit.

Reply:
I think they should be thread safe, I did not find in check results tsan report about them, though may be I missed.
I see here https://s3.amazonaws.com/clickhouse-test-reports/70823/1737e49046365dd7690f944463565b790f9baecf/integration_tests__tsan__[2_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log, the race is in line `auto filename_it = unique_filenames.emplace(filename).first;` (`MetadataStorageFromPlainRewritableObjectStorage.cpp:84`)